File: Java/src/main/java/parrot/Parrot.java
package parrot;

public interface Parrot {
    static Parrot createParrot(ParrotTypeEnum type, int numberOfCoconuts, double voltage, boolean isNailed) {
        return switch (type) {
            case EUROPEAN -> new EuropeanParrot();
            case AFRICAN -> new AfricanParrot(numberOfCoconuts);
            case NORWEGIAN_BLUE -> new NorwegianBlueParrot(voltage, isNailed);
        };
    }

    default double getBaseSpeed() {
        return 12.0;
    }

    double getSpeed();
    String getCry();
}

File: Scala/src/main/scala/parrot/Parrot.scala
package parrot

import parrot.ParrotType._

trait Parrot {
  def getSpeed: Double
  protected def getBaseSpeed: Double = 12.0
}

object Parrot {
  def createParrot(parrotType: ParrotType, numberOfCoconuts: Int, voltage: Double, isNailed: Boolean): Parrot = {
    parrotType match {
      case ParrotType.EUROPEAN => new EuropeanParrot
      case ParrotType.AFRICAN => new AfricanParrot(numberOfCoconuts)
      case ParrotType.NORWEGIAN_BLUE => new NorwegianBlueParrot(voltage, isNailed)
    }
  }
}
